src/envoy/http/error_translator/filter.cc

```diff
@@ -0,0 +1,211 @@
+      }
+    } else {
+      error_.set_code(Envoy::Grpc::Status::WellKnownGrpcStatus::Internal);
+    }
+    if (headers.GrpcMessage() != nullptr) {
+      const std::string grpc_message =
+          std::string(headers.GrpcMessage()->value().getStringView());
+      error_.set_message(grpc_message);
+    }
+
+    // Store in filter state.
+    utils::setErrorFilterState(*encoder_callbacks_->streamInfo().filterState(),
+                               error_);
+
+    return FilterHeadersStatus::Continue;
+  }
+
+  // Now we are handling `sendLocalReply` from non-espv2 filters.
+  // We need to modify the response body and filter state.
+  // TODO(nareddyt): Do we also need to check the content type is "plain/text"?
+  ENVOY_LOG(debug,
+            "Translating HTTP non-espv2 (upstream envoy filter) error headers");
+
+  // Stop iteration because we will modify some headers based on the body
+  // (example: content length). Expect encodeData to unblock this.
+  headers_ = &headers;
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterDataStatus Filter::encodeData(Envoy::Buffer::Instance& body,
+                                    bool end_stream) {
+  if (isUpstreamResponse()) {
+    // Do not translate any response sent from the upstream.
+    return FilterDataStatus::Continue;
+  }
+
+  if (is_grpc_response_) {
+    // Do not translate body for gRPC. It is not used.
+    return FilterDataStatus::Continue;
+  }
+
+  if (isEspv2FilterError()) {
+    ENVOY_LOG(debug, "Translating HTTP espv2 error body");
+    GOOGLE_DCHECK(!error_.message().empty());
+    GOOGLE_DCHECK(!error_json_.empty());
+
+    if (!end_stream) {
+      // Keep buffering data until we have the full response.
+      // TODO(nareddyt): Is there a better way to discard the remaining data? We
+      // don't care about it.
+      return FilterDataStatus::StopIterationAndBuffer;
+    }
+
+    // Replace the body with the JSON error.
+    Envoy::Buffer::OwnedImpl json_body(error_json_);
+    body.move(json_body);
+    return FilterDataStatus::Continue;
+  }
+
+  // Now we are handling `sendLocalReply` from non-espv2 filters.
+  // We need to modify the response body and filter state.
+  ENVOY_LOG(debug,
+            "Translating HTTP non-espv2 (upstream envoy filter) error body");
+  GOOGLE_DCHECK_NE(headers_, nullptr);
+
+  if (!end_stream) {
+    // Keep buffering data until we have the full response.
+    return FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  // Craft the error. Translate to canonical status codes.
+  const unsigned int http_code =
+      encoder_callbacks_->streamInfo().responseCode().value_or(500);
+  error_.set_code(Envoy::Grpc::Utility::httpToGrpcStatus(http_code));
+
+  const std::string body_string = body.toString();
+  error_.set_message(body_string);
+
+  // Set in the filter state for the access log (SC Report).
+  utils::setErrorFilterState(*encoder_callbacks_->streamInfo().filterState(),
```

Comment:
we can use the original rpc.Status in the filterState.   Why do we create a new one here?

Reply:
This is the non-espv2 && non-grpc case. There is no `google.rpc.Status` in the filter state, we create one based on the HTTP body in `encodeData()` here.